socorro/external/pubsub/crashqueue.py

```diff
@@ -0,0 +1,251 @@
+            logger.debug(
+                "PUBSUB_EMULATOR_HOST detected, connecting to emulator: %s",
+                emulator,
+            )
+        self.publisher = PublisherClient(
+            # publish messages in batches
+            batch_settings=BatchSettings(max_messages=publish_max_messages),
+            # disable retry in favor of socorro's retry and set rpc timeout
+            publisher_options=PublisherOptions(retry=None, timeout=publish_timeout),
+        )
+
+        self.standard_topic_path = self.publisher.topic_path(
+            project_id, standard_topic_name
+        )
+        self.priority_topic_path = self.publisher.topic_path(
+            project_id, priority_topic_name
+        )
+        self.reprocessing_topic_path = self.publisher.topic_path(
+            project_id, reprocessing_topic_name
+        )
+        self.queue_to_topic_path = {
+            "standard": self.standard_topic_path,
+            "priority": self.priority_topic_path,
+            "reprocessing": self.reprocessing_topic_path,
+        }
+
+        self.subscriber = SubscriberClient()
+        self.standard_subscription_path = self.subscriber.subscription_path(
+            project_id, standard_subscription_name
+        )
+        self.priority_subscription_path = self.subscriber.subscription_path(
+            project_id, priority_subscription_name
+        )
+        self.reprocessing_subscription_path = self.subscriber.subscription_path(
+            project_id, reprocessing_subscription_name
+        )
+        # order matters here, and is checked in tests
+        self.queue_to_subscription_path = {
+            "priority": self.priority_subscription_path,
+            "standard": self.standard_subscription_path,
+            "reprocessing": self.reprocessing_subscription_path,
+        }
+
+        self.pull_max_messages = pull_max_messages
+        self.publish_max_messages = publish_max_messages
+
+    def ack_crash(self, subscription_path, ack_id):
+        """Acknowledges a crash
+
+        :arg subscription_path: the subscription path for the queue
+        :arg ack_id: the ack_id for the message to acknowledge
+
+        """
+        self.subscriber.acknowledge(subscription=subscription_path, ack_ids=[ack_id])
+        logger.debug("ack %s from %s", ack_id, subscription_path)
+
+    def __iter__(self):
+        """Return iterator over crash ids from AWS SQS.
+
+        Each returned crash is a ``(crash_id, {kwargs})`` tuple with
+        ``finished_func`` as the only key in ``kwargs``. The caller should call
+        ``finished_func`` when it's done processing the crash.
+
+        """
+        while True:
+            has_msgs = {}
+            for subscription_path in self.queue_to_subscription_path.values():
+                resp = self.subscriber.pull(
+                    subscription=subscription_path,
+                    max_messages=self.pull_max_messages,
+                )
```

Comment:
`return_immediately` is deprecated, and I didn't use it because i thought it was throwing an exception when used, but i realize now it was throwing a warning and warnings are configured to error.

> Maybe we should use `return_immediately=True` just for the priority and reprocessing subscriptions and `return_immediately=False` for the standard subscription?

yeah, this seems like the right way to go

Reply:
> Warning: setting this field to true is discouraged because it adversely impacts the performance of Pull operations. We recommend that users do not set this field.

actually, after some thought, i think we don't care about this, and should just use `return_immediately=True` for all queues, because we are not a high-throughput use case in pubsub terms. We aren't trying to deliver billions of messages or hundreds of terrabytes of data per day, and we can revisit this if performance proves an issue.

Reply:
I (finally) found where it says `returnImmediately` is deprecated in the REST API docs. I'm not sure what that means in those docs. Is it deprecated and going to be removed? When? Etc.

Maybe instead, we should set the timeout for the client to 0.0 and see if that has the same effect? We should notice pretty quickly in the processor output as to whether it's spending a lot of time blocked on subscription pulls.

Reply:
> I (finally) found where it says `returnImmediately` is deprecated in the REST API docs.
> 
> https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/pull
> 
> I'm not sure what that means in those docs. Is it deprecated and going to be removed?

I don't think there's any explicit contract for what it means, but at the very least they won't remove it from the v1 api, they would cut a new version. if they instead decided to change to throwing an error or not respecting the option, we would get an announcement in gcp's release notes. I expect that they would give us enough notice to react accordingly if that happens.

based on the field description for `return_immediately`, I suspect that it's just a performance concern so they don't get complaints about pubsub being slow when it's used, which is fine for us:

> Optional. If this field set to true, the system will respond immediately even if it there are no messages available to return in the subscriptions.pull response. Otherwise, the system may wait (for a bounded amount of time) until at least one message is available, rather than returning no messages. Warning: setting this field to true is discouraged because it adversely impacts the performance of subscriptions.pull operations. We recommend that users do not set this field.

> Maybe instead, we should set the timeout for the client to 0.0

we definitely can't use the timeout like that, i checked by passing timeout when pulling from actual pubsub and it just causes `google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded`.